src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs

```diff
@@ -745,6 +737,361 @@ void M()
+                Diagnostic(ErrorCode.ERR_ExternHasBody, "local9").WithArguments("local9()").WithLocation(20, 44),
+                // (23,45): error CS0179: 'local11()' cannot be extern and declare a body
+                //         [DllImport("a")] extern static void local11() { } // 10
+                Diagnostic(ErrorCode.ERR_ExternHasBody, "local11").WithArguments("local11()").WithLocation(23, 45),
+                // (24,44): error CS0179: 'local12()' cannot be extern and declare a body
+                //         [DllImport("a")] extern static int local12() => 0; // 11
+                Diagnostic(ErrorCode.ERR_ExternHasBody, "local12").WithArguments("local12()").WithLocation(24, 44));
+        }
+
+        [Fact]
+        public void LocalFunctionExtern_Generic()
+        {
+            var source = @"
+using System;
+using System.Runtime.InteropServices;
+
+class C
+{
+#pragma warning disable 8321 // Unreferenced local function
+
+    void M()
+    {
+        [DllImport(""a"")] extern static void local1();
+        [DllImport(""a"")] extern static void local2<T>(); // 1
+
+        void local3()
+        {
+            [DllImport(""a"")] extern static void local1();
+            [DllImport(""a"")] extern static void local2<T2>(); // 2
+        }
+
+        void local4<T4>()
+        {
+            [DllImport(""a"")] extern static void local1(); // 3
+            [DllImport(""a"")] extern static void local2<T2>(); // 4
+        }
+
+        Action a = () =>
+        {
+            [DllImport(""a"")] extern static void local1();
+            [DllImport(""a"")] extern static void local2<T>(); // 5
+
+            void local3()
+            {
+                [DllImport(""a"")] extern static void local1();
+                [DllImport(""a"")] extern static void local2<T2>(); // 6
+            }
+
+            void local4<T4>()
+            {
+                [DllImport(""a"")] extern static void local1(); // 7
+                [DllImport(""a"")] extern static void local2<T2>(); // 8
+            }
+        };
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (12,10): error CS7042: The DllImport attribute cannot be applied to a method that is generic or contained in a generic type.
+                //         [DllImport("a")] extern static void local2<T>(); // 1
+                Diagnostic(ErrorCode.ERR_DllImportOnGenericMethod, "DllImport").WithLocation(12, 10),
+                // (17,14): error CS7042: The DllImport attribute cannot be applied to a method that is generic or contained in a generic type.
+                //             [DllImport("a")] extern static void local2<T2>(); // 2
+                Diagnostic(ErrorCode.ERR_DllImportOnGenericMethod, "DllImport").WithLocation(17, 14),
+                // (22,14): error CS7042: The DllImport attribute cannot be applied to a method that is generic or contained in a generic type.
```

Comment:
> method that is generic or contained in a generic type [](start = 88, length = 54)

The local function associated with the error is neither.

Reply:
How about "The DllImport attribute cannot be applied to a method that is generic or contained in a generic type or generic local function."

Reply:
Perhaps "The DllImport attribute cannot be applied to a method that is generic or contained in a generic method or type."


---
In reply to: [356797048](https://github.com/dotnet/roslyn/pull/40011#discussion_r356797048) [](ancestors = 356797048)